packages/tools/src/tools/VolumeCroppingTool.ts

```diff
@@ -0,0 +1,1534 @@
+          ) {
+            state.point[POINTINDEX.Z] = newZCenter;
+            state.sphereSource.setCenter(state.point);
+            state.sphereActor.getProperty().setColor(state.color);
+            state.sphereSource.modified();
+          }
+        });
+      }
+      const volumeActor = viewport.getDefaultActor()?.actor;
+      const mapper = volumeActor.getMapper() as vtkVolumeMapper;
+      const clippingPlanes = mapper.getClippingPlanes();
+      clippingPlanes[PLANEINDEX.XMAX].setOrigin(planeXmax.getOrigin());
+      clippingPlanes[PLANEINDEX.YMAX].setOrigin(planeYmax.getOrigin());
+      clippingPlanes[PLANEINDEX.ZMAX].setOrigin(planeZmax.getOrigin());
+    }
+    if (
+      this.configuration.showHandles &&
+      this.configuration.showCornerSpheres
+    ) {
+      this._updateCornerSpheres(viewport);
+    }
+    // this._updateCornerSpheres(viewport);
+    viewport.render();
+  };
+
+  _prepareImageDataForPicking = (viewport) => {
+    const volumeActor = viewport.getDefaultActor()?.actor;
+    if (!volumeActor) {
+      return;
+    }
+    // Get the imageData from the volumeActor
+    const imageData = volumeActor.getMapper().getInputData();
+
+    if (!imageData) {
+      console.error('No imageData found in the volumeActor');
+      return null;
+    }
+
+    // Get the voxelManager from the imageData
+    const { voxelManager } = imageData.get('voxelManager');
+
+    if (!voxelManager) {
+      console.error('No voxelManager found in the imageData');
+      return imageData;
+    }
+
+    // Create a fake scalar object to expose the scalar data to VTK.js
+    const fakeScalars = {
+      getData: () => {
+        return voxelManager.getCompleteScalarDataArray();
+      },
+      getNumberOfComponents: () => voxelManager.numberOfComponents,
+      getDataType: () =>
+        voxelManager.getCompleteScalarDataArray().constructor.name,
+    };
+
+    // Set the point data to return the fakeScalars
+    imageData.setPointData({
+      getScalars: () => fakeScalars,
+    });
+  };
+
+  _onMouseDownSphere = (evt) => {
+    const element = evt.currentTarget;
+    const viewportsInfo = this._getViewportsInfo();
+    const [viewport3D] = viewportsInfo;
+
+    const renderingEngine = getRenderingEngine(viewport3D.renderingEngineId);
+    const viewport = renderingEngine.getViewport(viewport3D.viewportId);
+    const mouseCanvas = [evt.offsetX, evt.offsetY];
+    // Find the sphere under the mouse
+    for (let i = 0; i < this.sphereStates.length; ++i) {
+      const sphereCanvas = viewport.worldToCanvas(this.sphereStates[i].point);
+      const dist = Math.sqrt(
+        Math.pow(mouseCanvas[0] - sphereCanvas[0], 2) +
+          Math.pow(mouseCanvas[1] - sphereCanvas[1], 2)
+      );
+      if (dist < 20) {
+        // 20 pixels threshold
+        this.draggingSphereIndex = i;
+        element.style.cursor = 'grabbing';
+        return;
+      }
+    }
+    this.draggingSphereIndex = null;
+  };
+
+  _onMouseMoveSphere = (evt) => {
```

Comment:
This is a big function, if possible extract code into separate named functions or utils so the tool is not too big and hard to understand.